Added Fallback for audio url

amitojsingh opened the pull request:
- Added Fallback URL if downloaded audio file doesn’t exists or corrupted .

amitojsingh committed:
fix audio functionality

amitojsingh committed:
fix tests

amitojsingh committed:
fix deep scan bugs

amitojsingh committed:
Merge branch 'dev' into fix-functionality

amitojsingh committed:
fix: bugs by uttamdeep kaur

amitojsingh committed:
Merge branch 'fix-functionality' into fix-autoplay

amitojsingh committed:
fix: auto play issue with trackPlayer

amitojsingh committed:
added test suite for AudioTrackDialog

amitojsingh committed:
Merge branch 'dev' into fix-functionality

amitojsingh committed:
Merge branch 'fix-functionality' into fix-autoplay

amitojsingh committed:
Refactor AudioPlayer and AudioTrackDialog components

- Removed console.log statements from AudioPlayer to clean up the code.
- Updated prop types in AudioTrackDialog to include new properties: displayName, audioUrl, lyricsUrl, trackLengthSec, and trackSizeMB for better data structure and validation.
- Adjusted import path in AudioControlBar test file for consistency.

amitojsingh committed:
Merge branch 'dev' into fix-autoplay

amitojsingh committed:
Added Fallback for audio url

Gauravjeetsingh reviewed (commented):
@amitojsingh veerji, I need some context on how are we using the tracks, downloaded tracks and remote tracks? What is the remote url here? 

Give me a quick rundown of how the audio urls work in the app, and I'll be able to review this PR more clearly.

amitojsingh commented:
- The user’s tracks are first fetched from the remote API.
- If remote tracks are available, they are merged with any locally downloaded tracks, and the audio URLs are updated to point to the local file paths where available.
- If no remote tracks are returned, the player falls back to using only the locally downloaded tracks.
- The audioUrl field stores the remote URL when a local file is not available.
- Additionally, the original remote URL is stored separately so that, if the local file is missing, corrupted, or not fully downloaded, playback can safely fall back to streaming from the remote source.

```mermaid
flowchart TD
  A[Start: user opens audio screen] --> B[Fetch remote tracks from API]

  B -->|Remote tracks found| C[Load downloaded tracks from local storage]
  B -->|No remote tracks| D[Load downloaded tracks only]

  C --> E[Merge remote and local by track id]
  E --> F{Local file exists?}

  F -->|Yes| G[Set audioUrl to local file]
  F -->|No| H[Set audioUrl to remote url]

  D --> I{Any downloaded tracks?}
  I -->|Yes| J[Set audioUrl to local file]
  I -->|No| K[No tracks available]

  G --> L[Store remote url as backup]
  H --> L
  J --> L

  L --> M[Start playback using audioUrl]
```

amitojsingh committed:
Merge branch 'dev' into fix-fallback-audioURL

Gauravjeetsingh commented on src/ReaderScreen/components/AudioPlayer/hooks/useAudioManifest/index.js:
```diff
@@ -34,44 +35,76 @@ const useAudioManifest = (baniID) => {
-        };
-      });
-    }
+      const validatedDownloads = await Promise.all(
```
If even one of the `exists()` threw an error, the entire operation fails. Wrap each file existence check in try catch block to handle the error gracefully.

Gauravjeetsingh commented on src/ReaderScreen/components/AudioPlayer/hooks/useAudioManifest/index.js:
Shouldn't this data include the remoteURL, if it exists?

```js
    ...(remoteUrl && { remoteUrl }), // Store remote URL if available
```

amitojsingh committed:
refactor: enhance audio and lyrics existence checks in useAudioManifest

- Updated the logic to handle file existence checks for audio and lyrics more robustly, ensuring that errors during checks are caught and treated as missing files.
- Added remoteUrl to the audio manifest for better track information management.
